### PR TITLE
use a @LoadBalanced Oauth2RestTemplate if configured

### PR DESCRIPTION
**Pull request by spencergibb**
fixes gh-51

@dsyer please review

### CODE REVIEW

File: src/main/java/org/springframework/cloud/security/oauth2/client/OAuth2LoadBalancerClientAutoConfiguration.java [hunk 1]
```diff
@@ -35,23 +39,45 @@
...
+                    restTemplate.getInterceptors());
+            interceptors.add(loadBalancerInterceptor);
+            restTemplate.setInterceptors(interceptors);
+
+            return restTemplate;
+        }
+    }
+
+    @Configuration
+    @ConditionalOnProperty(value = "spring.oauth2.userInfo.loadBalanced", matchIfMissing = false)
```
Comment (spencergibb) [inline src/main/java/org/springframework/cloud/security/oauth2/client/OAuth2LoadBalancerClientAutoConfiguration.java]: property name?

File: src/main/java/org/springframework/cloud/security/oauth2/proxy/OAuth2ProxyAutoConfiguration.java [hunk 2]
```diff
@@ -43,14 +44,24 @@
...
-	// Inject the @Primary one because it knows how to refresh an expired token 
+	// Inject the @Primary one because it knows how to refresh an expired token
 	@Autowired(required = false)
 	private OAuth2RestTemplate restTemplate;
 
+	@Autowired(required = false)
+	@LoadBalanced
+	private OAuth2RestTemplate loadBalancedRestTemplate;
+
+	@Value("${spring.oauth2.userInfo.loadBalanced:false}")
```
Comment (spencergibb) [inline src/main/java/org/springframework/cloud/security/oauth2/proxy/OAuth2ProxyAutoConfiguration.java]: property name?

File: src/main/java/org/springframework/cloud/security/oauth2/proxy/OAuth2ProxyAutoConfiguration.java [hunk 3]
```diff
@@ -43,14 +44,24 @@
...
+	private boolean useLoadBalancedRestTemplate;
+
 	@Bean
 	public OAuth2TokenRelayFilter oauth2TokenRelayFilter() {
 		OAuth2TokenRelayFilter filter = new OAuth2TokenRelayFilter(properties);
-		if (restTemplate!=null) {
+		if (loadBalancedRestTemplate != null && useLoadBalancedRestTemplate) {
```
Comment (spencergibb) [inline src/main/java/org/springframework/cloud/security/oauth2/proxy/OAuth2ProxyAutoConfiguration.java]: Seems a little hackish, but it works for me in myfeed.

File: src/main/java/org/springframework/cloud/security/oauth2/client/OAuth2LoadBalancerClientAutoConfiguration.java [hunk 4]
```diff
@@ -35,22 +39,46 @@
...
 					restTemplate.getInterceptors());
 			interceptors.add(loadBalancerInterceptor);
 			restTemplate.setInterceptors(interceptors);
+
+			return restTemplate;
+		}
+	}
+
+	@Configuration
+	@ConditionalOnProperty(value = "spring.oauth2.userInfo.loadBalanced", matchIfMissing = false)
```
Comment (spencergibb) [inline src/main/java/org/springframework/cloud/security/oauth2/client/OAuth2LoadBalancerClientAutoConfiguration.java]: load balanced rest template defaults to false.

File: src/main/java/org/springframework/cloud/security/oauth2/proxy/OAuth2ProxyAutoConfiguration.java (hunk 2 above)
Comment (dsyer) [inline src/main/java/org/springframework/cloud/security/oauth2/proxy/OAuth2ProxyAutoConfiguration.java]: Should probably be in `spring.oauth2.resource` (where the `userInfoUri` is already), i.e. in `ResourceServerProperties`.

File: src/main/java/org/springframework/cloud/security/oauth2/client/OAuth2LoadBalancerClientAutoConfiguration.java [hunk 5]
```diff
@@ -35,22 +40,46 @@
...
 					restTemplate.getInterceptors());
 			interceptors.add(loadBalancerInterceptor);
 			restTemplate.setInterceptors(interceptors);
+
+			return restTemplate;
+		}
+	}
+
+	@Configuration
+	@ConditionalOnProperty(value = ResourceServerProperties.PREFIX + ".loadBalanced", matchIfMissing = false)
```
Comment (spencergibb) [inline src/main/java/org/springframework/cloud/security/oauth2/client/OAuth2LoadBalancerClientAutoConfiguration.java]: Updated to use `ResourceServerProperties.PREFIX` aka `spring.oauth2.resource`

**Comment by spencergibb** [comment]
closed via 03fdfb1d83163187eadfe758a92d993f6da2f1d1

**Pull request closed without merging**